lib/spear/connection.ex

```diff
@@ -115,22 +116,94 @@ defmodule Spear.Connection do
 
   defp stream_body_message(message, {:ok, state}, request_ref) do
-    {wire_data, _byte_size} = Request.to_wire_data(message)
-
-    stream_result =
-      Mint.HTTP.stream_request_body(
-        state.conn,
-        request_ref,
-        wire_data
-      )
-
-    case stream_result do
-      {:ok, conn} -> {:cont, {:ok, put_in(state.conn, conn)}}
-      {:error, conn, reason} -> {:halt, {:error, put_in(state.conn, conn), reason}}
+    {wire_data, byte_size} = Request.to_wire_data(message)
+    smallest_window = get_smallest_window(state.conn, request_ref)
+
+    with false <- byte_size > smallest_window,
+         {:ok, conn} <- Mint.HTTP.stream_request_body(state.conn, request_ref, wire_data) do
+      {:cont, {:ok, put_in(state.conn, conn)}}
+    else
+      _window_too_small? = true ->
+        recv_until_window_size_increase(state, smallest_window, request_ref)
+
+      {:error, conn, reason} ->
+        {:halt, {:error, put_in(state.conn, conn), reason}}
+    end
+  end
+
+  defp recv_until_window_size_increase(state, current_window, request_ref) do
+    params = {current_window, request_ref}
+
+    {:ok, state}
+    |> do_stage(:recv_responses, params)
+    |> do_stage(:check_window_size, params)
+    |> emit_stage_results()
+  end
+
+  defp do_stage({:error, state, reason}, _stage, _params), do: {:error, state, reason}
+
+  defp do_stage({:ok, state}, :recv_responses, {_current_window, _request_ref}) do
+    case receive_next_and_stream(state.conn) do
+      {:ok, conn, responses} ->
+        state =
+          put_in(state.conn, conn)
+          |> handle_responses(responses)
+
+        {:ok, state}
+
+      {:error, conn, reason} ->
+        state = put_in(state.conn, conn)
+
+        {:error, state, reason}
+    end
+  end
+
+  defp do_stage({:ok, state}, :check_window_size, {current_window, request_ref}) do
+    new_window = get_smallest_window(state.conn, request_ref)
+
+    if new_window > current_window do
+      {:ok, state}
+    else
+      # if the window has not gotten bigger, loop back to the prior stage
+      # in the pipeline: block and wait for more messages from the server
+      do_stage({:ok, state}, :recv_responses, {new_window, request_ref})
+    end
+  end
+
+  defp emit_stage_results({:ok, state}), do: {:cont, {:ok, state}}
+  defp emit_stage_results({:error, state, reason}), do: {:halt, {:error, state, reason}}
+
+  defp receive_next_and_stream(conn) do
+    # YARD allow customization of timeout?
+    receive do
+      message when Mint.HTTP.is_connection_message(conn, message) ->
+        Mint.HTTP.stream(conn, message)
+    after
+      1_000 ->
+        {:error, conn, :window_update_timeout}
+    end
+  end
+
+  defp get_smallest_window(conn, request_ref) do
+    min(
+      Mint.HTTP2.get_window_size(conn, :connection),
+      safe_get_request_window_size(conn, request_ref)
+    )
+  end
+
+  defp safe_get_request_window_size(conn, request_ref) do
+    if Map.has_key?(conn.ref_to_stream_id, request_ref) do
```

Comment:
it appears as though `Mint.HTTP.get_window_size/2` throws an argument when the request is not in the `conn` (which it may be if the EventStore cancels our request from the server-side)

this thing I'm doing here with `conn.ref_to_stream_id` is how `Mint.HTTP.get_window_size/2` is implemented for the `{:request, request_ref}` input

kinda dirty to be searching a private field in a mint conn like this but /shrug didn't see anything in the public API to allow me to check if a request currently belongs to a conn

might be a good PR to mint